### PR TITLE
password security validation function created

### LINKED ISSUE #22: Add password security checks
## Description
As a user,
I need password security validation,
so that I can be sure that my password is secure.

Add validations that make the user choose strong passwords instead of insecure ones. This validation can be done here and on the frontend.

### Update
- [ ] Add password check to Change Password API
- [ ] Add password check to Register API
- [ ] Create password security validation function

## Definition of Done
- [ ] All of the required items are completed.
- [x] Approval by 1 mentor.

## Estimation
1 hours

### PR DESCRIPTION
**Pull request by chahat99**
### Description
Passwod security validation function has been created and now checks in register and change password api.

Fixes #22

### Type of Change:
- Code

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
The function is working fine locally.

### Checklist:
**Delete irrelevant options.**

- [ ] My PR follows the style guidelines of this project
- [ ] I have performed a self-review of my own code or materials

**Code/Quality Assurance Only**
- [ ] My changes generate no new warnings

### CODE REVIEW

**Comment by chahat99** [comment]
@isabelcosta I have created the password security validation function. Can you please review this?

File: app/messages.py [hunk 1]
```diff
@@ -213,3 +213,9 @@
 # Miscellaneous
 VALIDATION_ERROR = {"message": "Validation error."}
 NOT_IMPLEMENTED = {"message": "Not implemented."}
+
+#password errors
+INSUFFICIENT_LENGTH = {"message":"Password should have minimum 8 characters"}
+ATLEAST_ONE_UPPERCASE = {"message":"Password should have atleast 1 uppercase character"}
+ATLEAST_NUMBER = {"message":"Password should have atleast 1 number"}
+ATLEAST_ONE_SPECIAL_CHARACTER = {"message":"Password should have atleast 1 special character."}
```
Comment (sara-02) [inline app/messages.py]: Add space between "message":<space>" password...."

File: app/api/dao/user.py [hunk 2]
```diff
@@ -9,6 +9,36 @@
...
+        if(regex.search(password_check) is None):
+            return messages.ATLEAST_ONE_SPECIAL_CHARACTER,400
+        else:
+            flag=3
+    if flag==3:
+        return True
+    else:
+        return False
+
+
```
Comment (sara-02) [inline app/api/dao/user.py]: Hey.
Two things here:
1. You need to add spaces between variables. Basically run your code peace via some python linter/formatter.
2. Can we come up with a more efficient way of this algorithm rather than running the loop multiple times? I was thinking of using a flags dict and updating the count there for each type of flag. That way we will run the loop only twice. Once on the password text and once on the flag dict.
    Using a list of flags/bitwise operation can also be another way.
WDYT?

**Review by sara-02 — CHANGES_REQUESTED** [review]

File: app/api/dao/user.py (hunk 2 above)
Comment (chahat99) [reply app/api/dao/user.py]: Hi @sara-02
I am really sorry for the late reply.Actually I was down with fever and wasn't able to work since few days. I will surely format the code by adding the spaces.
Using dictionary is a good idea.According to me,there will only be one loop in that case changing the flag according to the password text.But we need to put a break like here so that all letters of the password are not checked once there is one mistake. Shall I modify the code using dict?

Comment (sara-02) [reply app/api/dao/user.py]: Hey, we can put a break, but in the current implementation, we are scanning the string again and again for each flag. That is why I thought if we can reduce the number of loops there. Also, what if tomorrow we add a new flag, can our code to flexible to accommodate those changes? A dict of flags or even a bitwise operation on a flag(assuming eight types of flags at a time) might be easier to handle than adding/deleting loops for each change.

Comment (sara-02) [reply app/api/dao/user.py]: Let me know if you have any more ideas.

Comment (chahat99) [reply app/api/dao/user.py]: What will the eight different types of flags include?

Comment (sara-02) [reply app/api/dao/user.py]: I was just floating an idea, the flags will include the same current set of caps, number, special char. I know it is not 8, but giving an idea that we can have bitwise flags as well.

Comment (chahat99) [reply app/api/dao/user.py]: Okay I will do the same

Comment (chahat99) [reply app/api/dao/user.py]: @sara-02 I've commited the changes you mentioned.Could you please review them?

**Comment by mtreacy002** [comment]
Hi @chahat99, @sara-02, @isabelcosta. 

Looking at @IlyaMaier suggestion, his PR [#423](https://github.com/anitab-org/mentorship-android/pull/423) that also addresses this issue [#22](https://github.com/anitab-org/mentorship-backend/issues/22) has valid point. If PR 423 on Android gets approved, password validation will be dealt at front-end and we don't need to do anything else for validating password at back-end. 

We probably only need to make few changes at back-end (like removing password hashing from user models and create test case for user registration/login using SHA256 as dummy password) if PR [#406](https://github.com/anitab-org/mentorship-android/pull/406) on Android that encrypt password also gets approved.

I like the idea of saving overhead on server-side but also aware that it might make testing backend a bit more tricky. Again, I'm a newbie offering feedback. I'm eager to learn how this issue gets resolved. 😊

File: app/api/dao/user.py [hunk 3]
```diff
@@ -9,6 +10,45 @@
...
+        'len': messages.INSUFFICIENT_LENGTH,
+        'uppcas': messages.ATLEAST_ONE_UPPERCASE,
+        'digits': messages.ATLEAST_NUMBER,
+        'specchar': messages.ATLEAST_ONE_SPECIAL_CHARACTER,
+        }
+
+    if len(password_check) < 8:
+        flags['len'] += 1
+    else:
+        for ch in password_check:
```
Comment (annabauza) [inline app/api/dao/user.py:45]: lets add lambda function for that maybe `count_chars = lambda str_input, str_chars : sum(str_input.count(c) for c in str_chars)` so then we can call things like `count_chars(test_str, string.digits)` python string has many predefined arrays, please check documentation

File: app/api/dao/user.py [hunk 4]
```diff
@@ -54,17 +97,19 @@ def create_user(data):
 
         return messages.USER_WAS_CREATED_SUCCESSFULLY, 200
 
+    # checks if the password is secure (Password security validation function)
+
     @staticmethod
     @email_verification_required
     def delete_user(user_id):
         """ Deletes a user.
-
```
Comment (annabauza) [inline app/api/dao/user.py:71]: can you change your on save lint so it will not doing unnecesary changes?

File: app/api/dao/user.py [hunk 5]
```diff
@@ -189,35 +234,65 @@ def update_user_profile(user_id, data):
...
-        if 'interests' in data and data['interests']:
-            user.interests = data['interests']
-
-        if 'resume_url' in data and data['resume_url']:
-            user.resume_url = data['resume_url']
-
-        if 'photo_url' in data and data['photo_url']:
-            user.photo_url = data['photo_url']
+        if 'bio' in data:
```
Comment (annabauza) [inline app/api/dao/user.py]: is this change related to this PR?

File: app/api/dao/user.py [hunk 6]
```diff
@@ -233,41 +308,42 @@ def update_user_profile(user_id, data):
...
 
         current_password = data['current_password']
         new_password = data['new_password']
 
         user = UserModel.find_by_id(user_id)
         if user.check_password(current_password):
-            user.set_password(new_password)
-            user.save_to_db()
-            return messages.PASSWORD_SUCCESSFULLY_UPDATED, 201
+            if check_security(new_password):
```
Comment (annabauza) [inline app/api/dao/user.py:359]: can we just use && in 1 if?

File: app/messages.py [hunk 7]
```diff
@@ -131,7 +131,19 @@
                              " not in the accepted state."}
 MENTORSHIP_RELATION_NOT_IN_ACCEPT_STATE = {"message": "Mentorship relation is"
                                            " not in the accepted state."}
-
+#password errors
+INSUFFICIENT_LENGTH = {
+  "message": "Password should have minimum 8 characters"
+}
+ATLEAST_ONE_UPPERCASE = {
+  "message": "Password should have atleast 1 uppercase character"
```
Comment (annabauza) [inline app/messages.py:179]: at least

File: app/messages.py [hunk 8]
```diff
@@ -131,7 +131,19 @@
...
+}
+ATLEAST_NUMBER = {
+  "message": "Password should have atleast 1 number"
```
Comment (annabauza) [inline app/messages.py:182]: at least

File: app/messages.py [hunk 9]
```diff
@@ -131,7 +131,19 @@
...
+}
+ATLEAST_ONE_SPECIAL_CHARACTER = {
+  "message": "Password should have atleast 1 special character."
```
Comment (annabauza) [inline app/messages.py:185]: at least

**Review by annabauza — CHANGES_REQUESTED** [review]
Please see my comments

**Comment by annabauza** [comment]
And you have conflicts to resolve

File: app/api/dao/user.py (hunk 3 above)
Comment (chahat99) [reply app/api/dao/user.py:45]: Okay you want to make a separate function for counting occurrences of a particular character in a string.I will check if there are any predefined ways in python.

File: app/api/dao/user.py (hunk 4 above)
Comment (chahat99) [reply app/api/dao/user.py:71]: @annabauza 
Okay I will change that.

File: app/api/dao/user.py (hunk 5 above)
Comment (chahat99) [reply app/api/dao/user.py]: @annabauza No this change is not related to this PR.I didn't do this change.

File: app/api/dao/user.py (hunk 6 above)
Comment (chahat99) [reply app/api/dao/user.py:359]: @annabauza where should we use &&? I didn't get that.

**Comment by devkapilbansal** [comment]
Closing as the issue seems to be already closed by a PR
